src/Cardano/Wallet/Kernel/DB/HdWallet.hs

```
@@ -569,10 +573,25 @@ mkAddressPool
     -> Core.PublicKey
     -> AddressPoolGap
     -> AddressPool Core.Address
-mkAddressPool mkAddress accPK gap = emptyAddressPool gap newAddress
-  where
-    newAddress :: Word32 -> Core.Address
-    newAddress addrIx = case deriveAddressPublicKey accPK ExternalChain addrIx of
+mkAddressPool mkAddress accPK gap
+    = emptyAddressPool gap (mkAddressBuilder mkAddress accPK)
+
+mkAddressPoolExisting
+    :: (Core.PublicKey -> Core.Address)
+    -> Core.PublicKey
+    -> AddressPoolGap
+    -> [(Core.Address, Word32)]
+    -> Either ErrAddressPoolInvalid (AddressPool Core.Address)
+mkAddressPoolExisting mkAddress accPK gap addrs
+    = initAddressPool gap (mkAddressBuilder mkAddress accPK) addrs
+
+mkAddressBuilder
+    :: (Core.PublicKey -> Core.Address)
+    -> Core.PublicKey
+    -> Word32
+    -> Core.Address
+mkAddressBuilder mkAddress accPK addrIx
+    = case deriveAddressPublicKey accPK ExternalChain addrIx of
         Nothing     -> error "mkAddressPool: maximum number of addresses reached."
```

We decided to keep `error` here, right? I mean, is it appropriate solution here?

---

I'll make that part of ErrAddressPoolInvalid instead...

---

(will remove error call)

---

@denisshevchenko on further thought, this behaviour is correct: this function will be deferred to during prefiltering (when an address pool gets extended). In that case it will have to fail hard if address generation fails. There are some other hard fails that occur during SwitchToFork and Appyblock